subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/SourceSetUtil.java

```diff
@@ -70,12 +76,26 @@ public File call() {
             }
         }));
 
+        final String annotationProcessorGeneratedSourcesChildPath = "generated/sources/annotationProcessor/" + sourceDirectorySet.getName() + "/" + sourceSet.getName();
+        sourceDirectorySet.setAnnotationProcessorGeneratedSourcesDirectory(target.provider(new Callable<File>() {
+            @Override
+            public File call() {
+                return new File(target.getBuildDir(), annotationProcessorGeneratedSourcesChildPath);
+            }
+        }));
+
         DefaultSourceSetOutput sourceSetOutput = Cast.cast(DefaultSourceSetOutput.class, sourceSet.getOutput());
         sourceSetOutput.addClassesDir(new Callable<File>() {
             @Override
             public File call() {
                 return sourceDirectorySet.getOutputDir();
             }
         });
+        sourceSetOutput.getAnnotationProcessorGeneratedSourcesDirs().from(new Callable<File>() {
```

Comment:
Here we should instead do a from(compileTask.options.generatedSourcesDir) and leave out the SourceDirectorySet middle man.

Also we should deprecate the three separate methods on `CompileOptions` and replace them with a single `DirectoryProperty getGeneratedSourcesDir`. This would allow us to track the location and the dependency on the task in one object. Right now the code here only tracks the location, but doesn't track that the task needs to be run. I know this is covered by the very broad `dependsOn(sourceSet.outputs)` right now, but we'll wanna get away from that at some point :)